### PR TITLE
DRAFT: First pass at geoip support in benthos

### PR DESCRIPTION
**Pull request by jamesharr**

### CODE REVIEW

File: internal/bloblang/query/methods_geoip.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	"github.com/oschwald/geoip2-golang"
+)
+
+var _ = registerSimpleMethod(
+	NewMethodSpec(
+		"geoip_city", "",
+	).Param(ParamString("path", "Path to city mmdb file")),
+	func(args *ParsedParams) (simpleMethod, error) {
+		// TODO: Ensure path is a static. Unsure how to accomplish this
```
Comment (Jeffail) [inline internal/bloblang/query/methods_geoip.go:14]: Leave this one with me, I'll need to add a way to optionally memoize constructors for functions like this, and part of that will likely be destruction as well (in which we'll call `db.Close()`).

File: internal/bloblang/query/methods_geoip.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+			case string:
+				ip = net.ParseIP(t)
+			default:
+				return nil, NewTypeError(v, ValueString)
+			}
+
+			// TODO: ip will be 'nil' if not a valid v4/v6 address. How to handle?
+			// It currently winds up reporting this when invalid:
+			//    "failed assignment (line 2): field `this.ip`: IP passed to Lookup cannot be nil"
+			// should we add context? IE what the lookup value is?
```
Comment (Jeffail) [inline internal/bloblang/query/methods_geoip.go:42]: If you return a bit of context like `fmt.Errorf("value %v does not appear to be a valid v4 or v6 IP address", v)` then it'll be wrapped with the extra context, so the resulting error should be something like `(line x): field this.ip: value foobar does not appear...`, I'd say that's fine for now

**Review by Jeffail — COMMENTED** [review]
Hey @jamesharr, looking good. I'd say the best approach would be to use something like https://github.com/mitchellh/mapstructure to convert the structured result into a generic `map[string]interface{}` form, at which point you'll be able to access those individual fields within bloblang.

In an ideal world the resulting structure would be snake case. I think we can likely do something like walk the tree of `map[string]interface{}` and rename all the keys by regular expressions without noticing a performance hit, but it maybe ought to be benchmarked. However, that also sounds like a drag so if you want to copy it through with camel case for now I'm happy to dig into that part myself.

**Comment by helgeolav** [comment]
I have not looked into the code yet, but support for geoenrichment is something I need. I, however, need to use ip2location as my provider.

When working on this, will it be possible to make it generic in code so Maxmind, ip2location and others can be added with more common code?

File: internal/bloblang/query/methods_geoip.go (hunk 1 above)
Comment (gallypette) [reply internal/bloblang/query/methods_geoip.go:14]: It looks like this is done. 
stracing this PR but merged with df919d7e9c7c3689132a5cf80799d3b5d0290308 shows that the mmdb file is only read once.

Comment (Jeffail) [reply internal/bloblang/query/methods_geoip.go:14]: I think it's good enough for now to proceed, ideally I want to add some sort of decaying cache but I don't think we need to block this PR on that.

Comment (gallypette) [reply internal/bloblang/query/methods_geoip.go:14]: Decaying would be nice but is really unnecessary for this use case imho.
Is the plan to wrap each geoip2 method into a bloblang method?

Comment (gallypette) [reply internal/bloblang/query/methods_geoip.go:14]: I am ok with it, it's just to be sure how to implement the rest ;)

**Comment by Jeffail** [comment]
Merged from https://github.com/Jeffail/benthos/pull/1068

**Pull request closed without merging**